### PR TITLE
Pre-warm intra block cache during block execution

### PR DESCRIPTION
**Pull request by LukaszRozmej**
## Changes

- During block processing, run the pending set of transactions in parallel using available processors to populate the read cache; for the sequential run of transactions
- Results in up to 2x performance when processing blocks. 

#### What types of changes does your code introduce?

- [x] New feature (a non-breaking change that adds functionality)
- [x] Optimization

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [ ] Yes
- [x] No

#### Notes on testing

Test eth_call and tracing.
Test on blocks that have self-destruct (archive sync?)

#### Requires documentation update

- [ ] Yes
- [x] No

#### Requires explanation in Release Notes

- [x] Yes
- [ ] No

### CODE REVIEW

File: src/Nethermind/Nethermind.Evm/VirtualMachine.cs [hunk 1]
```diff
@@ -181,6 +183,37 @@ public CallResult(ReadOnlyMemory<byte> output, bool? precompileSuccess, bool sho
     public interface IIsTracing { }
     public readonly struct NotTracing : IIsTracing { }
     public readonly struct IsTracing : IIsTracing { }
+
+    internal sealed class CodeLruCache
+    {
+        private const int CacheCount = 16;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/VirtualMachine.cs:189]: why 16 and not Environment.ProcessorCount for example?

File: src/Nethermind/Nethermind.Evm/VirtualMachine.cs [hunk 2]
```diff
@@ -181,6 +183,37 @@ public CallResult(ReadOnlyMemory<byte> output, bool? precompileSuccess, bool sho
...
+            return cache.Get(codeHash);
+        }
+
+        public bool Set(in ValueHash256 codeHash, CodeInfo codeInfo)
+        {
+            var cache = _caches[GetCacheIndex(codeHash)];
+            return cache.Set(codeHash, codeInfo);
+        }
+
+        private static int GetCacheIndex(in ValueHash256 codeHash) => codeHash.Bytes[^1] & CacheMax;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/VirtualMachine.cs:215]: why ^1 and not 0?

File: src/Nethermind/Nethermind.Evm/VirtualMachine.cs (hunk 1 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Evm/VirtualMachine.cs:189]: Distribution based on data; `x0-xf` in the nibble, 16 values

File: src/Nethermind/Nethermind.Evm/VirtualMachine.cs (hunk 2 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Evm/VirtualMachine.cs:215]: Disprotianate amount of contracts and addresses start with zeros; are gpu programs to hunt for the highest numbers of leading zeros due to calldata reductions:

https://github.com/johguse/ERADICATE2
https://github.com/1inch/profanity2

People don't search for the last nibbles so the distribution is fairly random

**Comment by benaadams** [comment]
Hive tests https://github.com/NethermindEth/nethermind/actions/runs/9228022573

**Comment by benaadams** [comment]
> Hive tests https://github.com/NethermindEth/nethermind/actions/runs/9228022573

Bunch of hive tests fail
```
Failed: 30

✖ Bad Hash on NewPayload (Syncing=true, Sidechain=true) (Paris) (nethermind)
✖ GetPayloadBodiesByHash (Empty Transactions/Withdrawals) (Paris) (nethermind)
✖ Inconsistent Finalized in ForkchoiceState (Paris) (nethermind)
✖ Inconsistent Safe in ForkchoiceState (Paris) (nethermind)
✖ Invalid Missing Ancestor ReOrg, StateRoot, EmptyTxs=True, Invalid P9 (Paris) (nethermind)
✖ Invalid Missing Ancestor Syncing ReOrg, ReceiptsRoot, EmptyTxs=False, CanonicalReOrg=True, Invalid P8 (Paris) (nethermind)
✖ Invalid Missing Ancestor Syncing ReOrg, StateRoot, EmptyTxs=False, CanonicalReOrg=True, Invalid P9 (Paris) (nethermind)
✖ Invalid Missing Ancestor Syncing ReOrg, StateRoot, EmptyTxs=True, CanonicalReOrg=True, Invalid P9 (Paris) (nethermind)
✖ Invalid NewPayload, PrevRandao, Syncing=False, EmptyTxs=False, DynFeeTxs=False (Paris) (nethermind)
✖ Invalid NewPayload, PrevRandao, Syncing=True, EmptyTxs=False, DynFeeTxs=False (Paris) (nethermind)
✖ Invalid NewPayload, ReceiptsRoot, Syncing=True, EmptyTxs=False, DynFeeTxs=False (Paris) (nethermind)
✖ Invalid NewPayload, StateRoot, Syncing=False, EmptyTxs=True, DynFeeTxs=False (Paris) (nethermind)
✖ Invalid NewPayload, Timestamp, Syncing=True, EmptyTxs=False, DynFeeTxs=False (Paris) (nethermind)
✖ Invalid NewPayload, Transaction ChainID, Syncing=True, EmptyTxs=False, DynFeeTxs=True (Paris) (nethermind)
✖ Invalid NewPayload, Transaction Gas, Syncing=True, EmptyTxs=False, DynFeeTxs=True (Paris) (nethermind)
✖ Invalid NewPayload, Transaction GasPrice, Syncing=True, EmptyTxs=False, DynFeeTxs=False (Paris) (nethermind)
✖ Invalid NewPayload, Transaction GasPrice, Syncing=True, EmptyTxs=False, DynFeeTxs=True (Paris) (nethermind)
✖ Invalid NewPayload, Transaction GasTipCapPrice, Syncing=True, EmptyTxs=False, DynFeeTxs=True (Paris) (nethermind)
✖ Invalid NewPayload, Transaction Nonce, Syncing=False, EmptyTxs=False, DynFeeTxs=True (Paris) (nethermind)
✖ Invalid NewPayload, Transaction Nonce, Syncing=True, EmptyTxs=False, DynFeeTxs=False (Paris) (nethermind)
✖ Invalid NewPayload, Transaction Nonce, Syncing=True, EmptyTxs=False, DynFeeTxs=True (Paris) (nethermind)
✖ Invalid NewPayload, Transaction Signature, Syncing=False, EmptyTxs=False, DynFeeTxs=False (Paris) (nethermind)
✖ Invalid NewPayload, Transaction Signature, Syncing=True, EmptyTxs=False, DynFeeTxs=False (Paris) (nethermind)
✖ Invalid NewPayload, Transaction Signature, Syncing=True, EmptyTxs=False, DynFeeTxs=True (Paris) (nethermind)
✖ Invalid PayloadAttributes, Zero timestamp, Syncing=False (Paris) (nethermind)
✖ Invalid PayloadAttributes, Zero timestamp, Syncing=True (Paris) (nethermind)
✖ Payload Build after New Invalid Payload: Invalid StateRoot (Paris) (nethermind)
✖ PrevRandao Opcode Transactions Test, DynamicFeeTransactions (Paris) (nethermind)
✖ Re-Org Back into Canonical Chain, Depth=10, Execute Side Payload on Re-Org (Paris) (nethermind)
✖ Sidechain Reorg (Paris) (nethermind)
```

**Comment by benaadams** [comment]
Hive tests https://github.com/NethermindEth/nethermind/actions/runs/9233024417

**Comment by benaadams** [comment]
```
Failed: 23

✖ GetPayloadBodiesByHash (Empty Transactions/Withdrawals) (Paris) (nethermind)
✖ Invalid Missing Ancestor ReOrg, StateRoot, EmptyTxs=True, Invalid P1 (Paris) (nethermind)
✖ Invalid Missing Ancestor ReOrg, StateRoot, EmptyTxs=True, Invalid P10 (Paris) (nethermind)
✖ Invalid Missing Ancestor ReOrg, StateRoot, EmptyTxs=True, Invalid P9 (Paris) (nethermind)
✖ Invalid Missing Ancestor Syncing ReOrg, GasLimit, EmptyTxs=False, CanonicalReOrg=False, Invalid P8 (Paris) (nethermind)
✖ Invalid Missing Ancestor Syncing ReOrg, GasLimit, EmptyTxs=False, CanonicalReOrg=True, Invalid P8 (Paris) (nethermind)
✖ Invalid Missing Ancestor Syncing ReOrg, GasUsed, EmptyTxs=False, CanonicalReOrg=True, Invalid P8 (Paris) (nethermind)
✖ Invalid Missing Ancestor Syncing ReOrg, StateRoot, EmptyTxs=False, CanonicalReOrg=True, Invalid P9 (Paris) (nethermind)
✖ Invalid Missing Ancestor Syncing ReOrg, StateRoot, EmptyTxs=True, CanonicalReOrg=True, Invalid P9 (Paris) (nethermind)
✖ Invalid Missing Ancestor Syncing ReOrg, Transaction GasPrice, EmptyTxs=False, CanonicalReOrg=True, Invalid P9 (Paris) (nethermind)
✖ Invalid NewPayload, Extra VersionedHashes, Syncing=True, EmptyTxs=False, DynFeeTxs=False (Cancun) (nethermind)
✖ Invalid NewPayload, GasLimit, Syncing=False, EmptyTxs=False, DynFeeTxs=False (Paris) (nethermind)
✖ Invalid NewPayload, GasLimit, Syncing=True, EmptyTxs=False, DynFeeTxs=False (Paris) (nethermind)
✖ Invalid NewPayload, Incomplete VersionedHashes, Syncing=False, EmptyTxs=False, DynFeeTxs=False (Cancun) (nethermind)
✖ Invalid NewPayload, Number, Syncing=True, EmptyTxs=False, DynFeeTxs=False (Paris) (nethermind)
✖ Invalid NewPayload, ParentHash, Syncing=True, EmptyTxs=False, DynFeeTxs=False (Paris) (nethermind)
✖ Invalid NewPayload, Transaction Gas, Syncing=True, EmptyTxs=False, DynFeeTxs=True (Paris) (nethermind)
✖ Invalid NewPayload, Transaction Nonce, Syncing=True, EmptyTxs=False, DynFeeTxs=True (Paris) (nethermind)
✖ Invalid NewPayload, Transaction Value, Syncing=True, EmptyTxs=False, DynFeeTxs=True (Paris) (nethermind)
✖ Invalid PayloadAttributes, Parent timestamp, Syncing=False (Paris) (nethermind)
✖ Re-Org Back into Canonical Chain, Depth=5 (Paris) (nethermind)
✖ Re-Org Back to Canonical Chain From Syncing Chain (Paris) (nethermind)
✖ Transaction Re-Org, Re-Org Out (Paris) (nethermind)
```

**Comment by benaadams** [comment]
Hive tests pass https://github.com/NethermindEth/nethermind/actions/runs/9234113726

**Comment by benaadams** [comment]
Consensus hive failures are recent EIP7610 tests https://github.com/ethereum/EIPs/pull/8161

**Review by asdacap — APPROVED** [review]
Not sure about passing in `PreBlockCaches` in `CreateResettableWorldState`. But it should work fine.

**Comment by LukaszRozmej** [comment]
> Not sure about passing in `PreBlockCaches` in `CreateResettableWorldState`. But it should work fine.

I took big concern passing it there, but it is easy way I found to support default behavior without those caches and then support behavior with caches when needed.

**Pull request merged**